src/llmcompressor/pipelines/sequential/helpers.py

```diff
@@ -0,0 +1,365 @@
+        code = self.graph.python_code("self")
+        exec(code.src, code.globals)
+        return code.globals.get("forward")
+
+
+def trace_subgraphs(
+    model: PreTrainedModel,
+    sample_input: Dict[str, Any],
+    sequential_targets: List[str],
+    ignore: List[str],
+) -> List[Subgraph]:
+    """
+    Trace a model to produce subgraphs, where each sequential target belongs to exactly
+    one subgraph and where executing each subgraph in order is equivalent to executing
+    the original model
+
+    :param model: model being traced
+    :param sample_input: inputs whose values will change during execution but whose
+        __len__, __bool__, and __contains__ values are assumed constant across batches
+    :param sequential_targets: list of patterns matching sequential targets
+    :param ignore: list of patterns matching modules to ignore during tracing
+    :return: a list of Subgraphs in order of execution
+    """
+    # find modules
+    sequential_targets = match_modules(model, sequential_targets)
+    ignore = match_modules(model, ignore)
+
+    # initialize arguments
+    tracer = get_tracer(model, sequential_targets, ignore)
+    concrete_args = populate_concrete_args(model, sample_input)
+
+    # trace
+    with (
+        calibration_forward_context(model),
+        HooksMixin.disable_hooks(),
+    ):
+        graph = GraphModule(
+            model,
+            tracer.trace(
+                model,
+                dummy_inputs=sample_input,
+                concrete_args=concrete_args,
+                complete_concrete_args_with_inputs_not_in_dummy_inputs=False,
+                # bug in trace throws an error for variadic
+                # args and kwargs in function signature
+            ),
+        )
+
+    # copy metadata
+    graph.config = model.config
+    graph.class_for_deserialization = model.__class__
+    graph.device = model.device
+
+    # perform subgraph partition
+    partitions = topological_partition(graph, sequential_targets)
+    subgraphs = partition_graph(model, partitions)
+    trace_consumed_names(subgraphs)
+
+    return subgraphs
+
+
+def get_tracer(
+    model: Module, sequential_targets: Set[Module], ignore: Set[Module]
+) -> HFTracer:
+    """
+    Get a tracer specialized for the given model. The resulting tracer will not trace
+    inside of sequential targets, ignored targets, or offloaded modules.
+
+    Tracing within sequential targets and ignored targets is unnecessary, and tracing
+    within offloaded modules may result in meta tensors being added to the model graph
+
+    :param model: model being traced
+    :param sequential_targets: modules which are sequential targets
+    :param ignore: modules which are ignored
+    """
+    offloaded_modules = set(m for m in model.modules() if has_offloaded_params(m))
+    skip_trace_modules = sequential_targets | offloaded_modules | ignore
+
+    class SequentialTracer(HFTracer):
+        def create_arg(self, a: Any) -> Argument:
```

Comment:
can you explain why this is needed?

Reply:
> special extension allows models which depend on config values to be traced

I overload this function to insert my own definition for creating an argument which is of type `PretrainedConfig`. Many models use values from the config during execution, but `torch.fx` is only capable of "baking" a limited set of class types into the model graph.

This code says, whenever the graph would try to reference an instance of a `PretrainedConfig` (for example, to get an attribute like `config.max_sequence_length`), instead just create a `PretrainedConfig` on the fly and initialize it with all of the args from the original config.

Reply:
There are a few things in this file that we should consider upstreaming HF, this might be one of them